### PR TITLE
Deprecate legacy items

### PR DESCRIPTION
**Pull request by wprzytula**
# The following items are now marked as deprecated:

### Serialization:
- `scylla::impl_serialize_value_via_value`
- `scylla::impl_serialize_row_via_value_list`
- `scylla::ValueList`
- `scylla::serialize::batch::LegacyBatchValuesAdapter`
- `scylla::serialize::batch::LegacyBatchValuesIteratorAdapter`
- `scylla::serialize::row::ValueListAdapter`
- `scylla::serialize::row::ValueListToSerializeRowAdapterError`
- `scylla::serialize::row::serialize_legacy_row`
- `scylla::serialize::value::ValueAdapter`
- `scylla::serialize::value::ValueToSerializeValueAdapterError`
- `scylla::serialize::value::serialize_legacy_value`
- `scylla_cql::frame::value::LegacyBatchValues`
- `scylla_cql::frame::value::LegacyBatchValuesFirstSerialized`
- `scylla_cql::frame::value::LegacyBatchValuesIterator`
- `scylla_cql::frame::value::LegacyBatchValuesFromIter`
- `scylla_cql::frame::value::LegacyBatchValuesIteratorFromIterator`
- `scylla_cql::frame::value::TupleValuesIter`
- `scylla_cql::frame::value::Value`

### Deserialization:
- `scylla::impl_from_cql_value_from_method`
- `scylla::transport::iterator::LegacyNextRowError`
- `scylla::transport::legacy_query_result::IntoLegacyQueryResultError`
- `scylla::transport::legacy_query_result::MaybeFirstRowTypedError`
- `scylla::transport::legacy_query_result::RowsExpectedError`
- `scylla::transport::legacy_query_result::SingleRowError`
- `scylla::transport::legacy_query_result::IntoTypedRows`
- `scylla::transport::legacy_query_result::TypedRowIter`
- `scylla_cql::cql_to_rust::FromCqlValError`
- `scylla_cql::cql_to_rust::FromRowError`

Fixes: #1139 

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~~[ ] I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Comment by wprzytula** [comment]
semver-checks deem this PR API-breaking, but if you read its message through, you'll see that it only requires a minor version bump, not major - so it's OK to do the changes in 0.15.1 (because we consider the last number the minor version number).

**Comment by Lorak-mmk** [comment]
> semver-checks deem this PR API-breaking, but if you read its message through, you'll see that it only requires a minor version bump, not major - so it's OK to do the changes in 0.15.1 (because we consider the last number the minor version number).

We should make message and tag different if only a minor version bump is required. Especially after 1.0

**Review by Lorak-mmk — DISMISSED** [review]

**Review by muzarski — APPROVED** [review]

**Comment by wprzytula** [comment]
Please don't merge yet. I'll update thiserror after 2.0.6 got released, so that there are no unnecessary module-level `#![allow(deprecated)] attributes.

File: scylla-cql/src/frame/response/cql_to_rust.rs [hunk 1]
```diff
@@ -10,6 +10,11 @@ use std::net::IpAddr;
...
 use uuid::Uuid;
 
+#[deprecated(
+    since = "0.15.1",
+    note = "Legacy deserialization API is inefficient and is going to be removed soon"
+)]
+#[allow(deprecated)]
 #[derive(Error, Debug, Clone, PartialEq, Eq)]
 pub enum FromRowError {
     #[error("{err} in the column with index {column}")]
```
Comment (Lorak-mmk) [inline scylla-cql/src/frame/response/cql_to_rust.rs:20]: Why do you need `#[allow(deprecated)]` on a `#[deprecated` item?

File: scylla/src/transport/errors.rs [hunk 2]
```diff
@@ -34,15 +34,15 @@ use thiserror::Error;
...
 
-use super::{
-    iterator::NextRowError,
-    legacy_query_result::IntoLegacyQueryResultError,
-    query_result::{IntoRowsResultError, SingleRowError},
-};
+use super::iterator::NextRowError;
+#[allow(deprecated)]
+use super::legacy_query_result::IntoLegacyQueryResultError;
+use super::query_result::{IntoRowsResultError, SingleRowError};
```
Comment (Lorak-mmk) [inline scylla/src/transport/errors.rs:40]: Importing a deprecated item requires `#[allow(deprecated)]`?

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]
Commit: "codewide: deprecate legacy serialization API's items "

Do we strictly need this "legacy" module? Why not just deprecate the items?
If it is required for some reason, then it's creation should be a separate commit, not the same one that adds deprecation.

File: scylla/src/transport/errors.rs (hunk 2 above)
Comment (wprzytula) [reply scylla/src/transport/errors.rs:40]: Indeed.

Comment (wprzytula) [reply scylla/src/transport/errors.rs:40]: You need that allow in every place where you name a deprecated item.

Comment (Lorak-mmk) [reply scylla/src/transport/errors.rs:40]: Ok

File: scylla-cql/src/frame/response/cql_to_rust.rs (hunk 1 above)
Comment (Lorak-mmk) [reply scylla-cql/src/frame/response/cql_to_rust.rs:20]: Ah, probably because one of the variants is deprecated?

**Comment by wprzytula** [comment]
> Do we strictly need this "legacy" module? Why not just deprecate the items?

There are a huge lot of items, it would be a lot of meaningless work to attribute each of them.
I will split the commit into two.

**Comment by Lorak-mmk** [comment]
> > Do we strictly need this "legacy" module? Why not just deprecate the items?
> 
> There are a huge lot of items, it would be a lot of meaningless work to attribute each of them. I will split the commit into two.

But you are also applying deprecation notices to the items inside this new module, not only to the module itself.
So what work was avoided by creating this module?

**Comment by wprzytula** [comment]
> > > Do we strictly need this "legacy" module? Why not just deprecate the items?
> > 
> > 
> > There are a huge lot of items, it would be a lot of meaningless work to attribute each of them. I will split the commit into two.
> 
> But you are also applying deprecation notices to the items inside this new module, not only to the module itself. So what work was avoided by creating this module?

Those notices from inside the new module were leftovers. Now there is per-module global deprecation notice.

File: scylla-cql/src/frame/value.rs [hunk 3]
```diff
@@ -26,6 +20,18 @@ pub struct ValueTooBig;
...
+    #[error("Too many values to add, max 65,535 values can be sent in a request")]
+    TooManyValues,
+    #[error("Mixing named and not named values is not allowed")]
+    MixingNamedAndNotNamedValues,
+    #[error(transparent)]
+    ValueTooBig(#[from] ValueTooBig),
+    #[error("Parsing serialized values failed")]
+    ParseError,
+}
+
```
Comment (Lorak-mmk) [inline scylla-cql/src/frame/value.rs:39]: Why is this not part of the legacy module?

Comment (Lorak-mmk) [reply scylla-cql/src/frame/value.rs:39]: ditto for `ValueTooBig`

Comment (wprzytula) [reply scylla-cql/src/frame/value.rs:39]: `SerializeValuesError` is used in `types::serialize::row::SerializedValues::add_value`, and `SerializeValuesError` contains a variant for `ValueTooBig`. That's why I left those not deprecated.

**Review by Lorak-mmk — APPROVED** [review]

**Review by muzarski — APPROVED** [review]

**Pull request merged**